### PR TITLE
Add support for dataverse file revisions

### PR DESCRIPTION
**Pull request by rliebz**

### CODE REVIEW

File: waterbutler/providers/dataverse/provider.py [hunk 1]
```diff
@@ -129,7 +133,17 @@ def metadata(self, path, state=None, **kwargs):
 
     @asyncio.coroutine
     def revisions(self, path, **kwargs):
-        raise exceptions.ProviderError({'message': 'Dataverse does not support file revisions.'}, code=405)
+        versions = ['latest', 'latest-published']
```
Comment (chrisseto) [inline waterbutler/providers/dataverse/provider.py]: Can this be defined as a class level tuple?

Comment (rliebz) [inline waterbutler/providers/dataverse/provider.py]: If we ever do decide to allow numbered versions (I'm not sure yet how viable that would be through the API), each dataset would have a different list of versions. That being said, for the immediate future, these will be the only two versions. I think this way is a bit more conducive to future changes.

**Comment by rliebz** [comment]
Requires https://github.com/CenterForOpenScience/osf.io/pull/2570

File: waterbutler/providers/dataverse/provider.py [hunk 2]
```diff
@@ -167,9 +229,17 @@ def _get_all_data(self):
         published_files = published_data if isinstance(published_data, list) else []
```
Comment (chrisseto) [inline waterbutler/providers/dataverse/provider.py]: What could `published_data` be other than a list?

File: waterbutler/providers/dataverse/provider.py [hunk 3]
```diff
@@ -159,7 +221,7 @@ def _get_data(self, version):
 
     @asyncio.coroutine
     def _get_all_data(self):
-        # Unspecified (file view page), check both sets for metadata
+        """Get list of file metadata for all dataset versions"""
         try:
             published_data = yield from self._get_data('latest-published')
         except exceptions.MetadataError:
```
Comment (chrisseto) [inline waterbutler/providers/dataverse/provider.py]: You may want to do some checks on the status code here, if dataverse is down and returns a 503 that should be reraised.

File: waterbutler/providers/dataverse/provider.py [hunk 4]
```diff
@@ -129,16 +168,39 @@ def metadata(self, path, state=None, **kwargs):
...
         :param str version:
-            'latest' for draft files
-            'latest-published' for published files
+
+            - 'latest' for draft files
+            - 'latest-published' for published files
+            - None for all data
         """
 
+        if not version:
```
Comment (chrisseto) [inline waterbutler/providers/dataverse/provider.py]: Should version default to `None`?

File: waterbutler/providers/dataverse/provider.py [hunk 5]
```diff
@@ -129,16 +168,39 @@ def metadata(self, path, state=None, **kwargs):
...
+            data = yield from self._get_all_data()
```
Comment (chrisseto) [inline waterbutler/providers/dataverse/provider.py]: You can also just `return (yield from self._get_all_data())`

File: waterbutler/providers/dataverse/provider.py [hunk 6]
```diff
@@ -129,16 +168,39 @@ def metadata(self, path, state=None, **kwargs):
...
+        """Get past versions of the request file. Currently only checks
+        'latest' and 'latest-published' versions.
+
+        :param str path: The path to a key
+        :rtype list:
+        """
+        versions = ['latest', 'latest-published']
+        revisions = []
+        for version in versions:
+            metadata = yield from self._get_data(version)
```
Comment (chrisseto) [inline waterbutler/providers/dataverse/provider.py]: Any reason not to us get_all_data here and then iterate over that?

File: waterbutler/providers/dataverse/metadata.py [hunk 7]
```diff
@@ -77,3 +77,18 @@ def serialized(self):
...
+    def version_identifier(self):
+        return 'version'
+
+    @property
+    def version(self):
+        return self.raw
+
+    @property
+    def modified(self):
+        return None
```
Comment (chrisseto) [inline waterbutler/providers/dataverse/metadata.py]: Please add a new line at the end of this file, you may want to make pycharm add an extra line by default.

**Comment by chrisseto** [comment]
first pass done :racehorse:

File: waterbutler/providers/dataverse/provider.py (hunk 2 above)
Comment (rliebz) [inline waterbutler/providers/dataverse/provider.py]: Fixed. This was an issue with the way dataset metadata was being serialized.

File: waterbutler/providers/dataverse/provider.py (hunk 3 above)
Comment (rliebz) [inline waterbutler/providers/dataverse/provider.py]: Done.

File: waterbutler/providers/dataverse/provider.py (hunk 4 above)
Comment (rliebz) [inline waterbutler/providers/dataverse/provider.py]: Done.

File: waterbutler/providers/dataverse/provider.py (hunk 5 above)
Comment (rliebz) [inline waterbutler/providers/dataverse/provider.py]: Neat syntax. Done.

File: waterbutler/providers/dataverse/provider.py (hunk 6 above)
Comment (rliebz) [inline waterbutler/providers/dataverse/provider.py]: The idea was to specifically iterate over possible versions, but now that I look at it, that should be done in `_get_all_data` anyway. Done.

File: waterbutler/providers/dataverse/metadata.py (hunk 7 above)
Comment (rliebz) [inline waterbutler/providers/dataverse/metadata.py]: Done and done.

**Comment by chrisseto** [comment]
If the corresponding osf PR is a hotfix this will need to be upgraded to a hotfix as well.

**Comment by rliebz** [comment]
Moving now.

**Pull request closed without merging**